File: packages/shared/src/diagnostic.ts
import type { HarnessWorkflow, HarnessStage, HarnessAttemptPhase, HarnessAttemptStatus, HarnessStatus, HarnessCommitStatus } from "./harness";

/** Diagnostic hints only; HTTP completion never establishes domain commit. */
export interface DiagnosticEventV1 {
  schema_version: "diagnostic-event-v1";
  event_id: string;
  source: "server" | "browser" | "desktop";
  name: "request_started" | "response_headers" | "request_finished" | "request_failed" | "request_cancelled" | "lifecycle_started" | "lifecycle_stopped" | "harness_reference" | "resource_reference" | "provider_started" | "provider_finished" | "provider_failed" | "provider_attempt_started" | "provider_attempt_finished" | "provider_attempt_failed" | "tool_started" | "tool_finished" | "tool_failed" | "tool_unknown" | "action_started" | "action_finished" | "action_failed" | "action_cancelled";
  action_name?: "json_import_read_persona" | "json_import_read_scene" | "json_export_handoff" | "settings_save" | "vault_create" | "vault_unlock" | "vault_lock" | "vault_load_secrets" | "vault_save_secrets" | "vault_clear_secrets" | null;
  span_id?: string | null;
  parent_span_id?: string | null;
  tool_metric?: DiagnosticToolMetricV1 | null;
  provider_metric?: DiagnosticProviderMetricV1 | null;
  resource?: { resource_type: "persona"; resource_id: string; revision: number | null } | null;
  harness?: { operation_id: string; workflow: HarnessWorkflow; stage: HarnessStage; trace_id: string | null; attempt_id?: string | null; attempt_index?: number | null; phase?: HarnessAttemptPhase | null; attempt_status?: HarnessAttemptStatus | null } | null;
  category: "transport" | "lifecycle" | "harness" | "resource" | "provider" | "tool" | "action";
  severity: "info" | "warning" | "error";
  outcome: "started" | "headers_received" | "completed" | "failed" | "cancelled" | "observed" | "unknown";
  error_code: "transport_failure" | "http_error" | "cancelled" | "provider_failure" | "tool_failure" | "action_failure" | null;
  timestamp: string;
  request_id: string | null;
  client_instance_id: string | null;
  page_view_id: string | null;
  flow_id: string | null;
  action_id: string | null;
  duration_ms: number | null;
  status_code: number | null;
  method: "GET" | "POST" | "PUT" | "PATCH" | "DELETE" | "OPTIONS" | "HEAD" | null;
  route: string | null;
}

export const DIAGNOSTIC_EVENT_CATALOG = {
  request_started: ["transport", "info", "started", null],
  response_headers: ["transport", "info", "headers_received", null],
  request_finished: ["transport", "info", "completed", null],
  request_failed: ["transport", "error", "failed", "transport_failure"],
  request_cancelled: ["transport", "warning", "cancelled", "cancelled"],
  lifecycle_started: ["lifecycle", "info", "started", null],
  lifecycle_stopped: ["lifecycle", "info", "completed", null],
  harness_reference: ["harness", "info", "observed", null],
  resource_reference: ["resource", "info", "observed", null],
  provider_started: ["provider", "info", "started", null],
  provider_finished: ["provider", "info", "completed", null],
  provider_failed: ["provider", "error", "failed", "provider_failure"],
  provider_attempt_started: ["provider", "info", "started", null],
  provider_attempt_finished: ["provider", "info", "completed", null],
  provider_attempt_failed: ["provider", "error", "failed", "provider_failure"],
  tool_started: ["tool", "info", "started", null],
  tool_finished: ["tool", "info", "completed", null],
  tool_failed: ["tool", "error", "failed", "tool_failure"],
  tool_unknown: ["tool", "warning", "unknown", null],
  action_started: ["action", "info", "started", null],
  action_finished: ["action", "info", "completed", null],
  action_failed: ["action", "error", "failed", "action_failure"],
  action_cancelled: ["action", "warning", "cancelled", "cancelled"],
} as const;

export function classifyDiagnostic(name: DiagnosticEventV1["name"], statusCode: number | null = null):
  Pick<DiagnosticEventV1, "category" | "severity" | "outcome" | "error_code"> {
  const [category, severity, outcome, error_code] = DIAGNOSTIC_EVENT_CATALOG[name];
  const result: Pick<DiagnosticEventV1, "category" | "severity" | "outcome" | "error_code"> = { category, severity, outcome, error_code };
  if ((name === "response_headers" || name === "request_finished") && statusCode !== null && statusCode >= 400) {
    result.severity = statusCode >= 500 ? "error" : "warning";
    result.error_code = "http_error";
    if (name === "request_finished") result.outcome = "failed";
  }
  return result;
}


/** Eventually consistent diagnostic projection. Resolve canonical records for commit truth. */
export interface DiagnosticHarnessIndexV1 {
  schema_version: "diagnostic-harness-index-v1";
  trace_id: string;
  parent_trace_id: string | null;
  operation_id: string | null;
  workflow: HarnessWorkflow | null;
  stage: HarnessStage | null;
  state: "prepared" | "claimed" | "terminal" | null;
  status: HarnessStatus | null;
  commit_status: HarnessCommitStatus | null;
  duration_ms: number | null;
  started_at: string | null;
  completed_at: string | null;
  source_updated_at: string | null;
  gap: "source_removed" | "source_invalid_or_unavailable" | "terminal_trace_not_available" | null;
  components: { name: string; version: string }[];
  attempts: { attempt_id: string; attempt_index: number; phase: HarnessAttemptPhase; status: HarnessAttemptStatus; duration_ms: number }[];
  provider: null; model: null; tokens: null; cost: null;
  usage_gap: "not_recorded_in_canonical_trace";
  correlation_gap: "resolve_request_links_separately";
}


export interface DiagnosticProviderMetricV1 {
  adapter: "litellm";
  adapter_contract: "diagnostic-provider-transport-v1";
  request_kind: "plan" | "chat" | "setting" | "embedding" | "other";
  model: string | null;
  model_gap: "model_label_unreviewed" | null;
  timeout_seconds: number;
  max_attempts: 3;
  attempt_index: number | null;
  attempts_used: number;
  recovered: boolean;
  input_tokens: number | null;
  output_tokens: number | null;
  total_tokens: number | null;
  usage_source: "provider_reported" | "unavailable";
  usage_gap: "not_returned" | "partial_or_invalid" | "aggregate_not_additive" | null;
  cost: null;
  cost_gap: "provider_cost_evidence_unavailable";
  configuration_gap: "endpoint_configuration_not_recorded";
}


export interface DiagnosticToolMetricV1 {
  workflow: HarnessWorkflow;
  offered_in_stage: HarnessStage;
  execution_stage: HarnessStage | null;
  manifest_key: string | null;
  canonical_name: string | null;
  input_contract_version: string | null;
  result_contract_version: string | null;
  provider_tool_call_id: string | null;
  max_calls_per_operation: number | null;
  max_calls_per_round: number | null;
  timeout_ms: number | null;
  manifest_gap: "unregistered_tool" | null;
  effect_commit_claim: "none";
}

File: services/ai/app/models/diagnostic.py
"""Content-free diagnostic hints; never authorization or commit evidence."""
from typing import Annotated, Literal
from app.models.diagnostic_desktop import DiagnosticDesktopMetricV1
from app.models.diagnostic_tool import DiagnosticToolMetricV1
from app.models.diagnostic_provider import DiagnosticProviderMetricV1
from pydantic import BaseModel, ConfigDict, Field, model_validator
from app.models.harness import HarnessStage, HarnessWorkflow, HarnessAttemptPhase, HarnessAttemptStatus
from app.models.harness_operation import HARNESS_OPERATION_ID_PATTERN

DiagnosticPagePath = Literal["/", "/plan", "/study", "/persona-spectrum", "/scene-setup", "/tavern", "/settings", "/sensory-tools", "/model-usage"]

Identity = Annotated[str, Field(pattern=r"^[a-zA-Z0-9_-]{1,96}$")]


DIAGNOSTIC_EVENT_CATALOG = {
    "decode_failed": ("decode", "error", "failed", "decode_failure"),
    "page_entered": ("page", "info", "started", None),
    "page_left": ("page", "info", "completed", None),
    "request_started": ("transport", "info", "started", None),
    "response_headers": ("transport", "info", "headers_received", None),
    "request_finished": ("transport", "info", "completed", None),
    "request_failed": ("transport", "error", "failed", "transport_failure"),
    "request_cancelled": ("transport", "warning", "cancelled", "cancelled"),
    "lifecycle_started": ("lifecycle", "info", "started", None),
    "lifecycle_stopped": ("lifecycle", "info", "completed", None),
    "harness_reference": ("harness", "info", "observed", None),
    "resource_reference": ("resource", "info", "observed", None),
    "provider_started": ("provider", "info", "started", None),
    "provider_finished": ("provider", "info", "completed", None),
    "provider_failed": ("provider", "error", "failed", "provider_failure"),
    "provider_attempt_started": ("provider", "info", "started", None),
    "provider_attempt_finished": ("provider", "info", "completed", None),
    "provider_attempt_failed": ("provider", "error", "failed", "provider_failure"),
    "tool_started": ("tool", "info", "started", None),
    "tool_finished": ("tool", "info", "completed", None),
    "tool_failed": ("tool", "error", "failed", "tool_failure"),
    "tool_unknown": ("tool", "warning", "unknown", None),
    "action_started": ("action", "info", "started", None),
    "action_finished": ("action", "info", "completed", None),
    "action_failed": ("action", "error", "failed", "action_failure"),
    "action_cancelled": ("action", "warning", "cancelled", "cancelled"),
    "desktop_started": ("desktop", "info", "started", None),
    "sidecar_spawned": ("desktop", "info", "started", None),
    "sidecar_ready": ("desktop", "info", "completed", None),
    "sidecar_startup_failed": ("desktop", "error", "failed", "desktop_failure"),
    "sidecar_exited": ("desktop", "error", "failed", "desktop_failure"),
    "desktop_shutdown_requested": ("desktop", "info", "started", None),
    "sidecar_stopped": ("desktop", "info", "completed", None),
    "sidecar_shutdown_unknown": ("desktop", "warning", "unknown", None),
    "desktop_stopped": ("desktop", "info", "completed", None),
}


def classify_diagnostic(name, status_code=None):
    category, severity, outcome, error_code = DIAGNOSTIC_EVENT_CATALOG[name]
    if name in {"response_headers", "request_finished"} and isinstance(status_code, int) and status_code >= 400:
        severity, error_code = "error" if status_code >= 500 else "warning", "http_error"
        if name == "request_finished":
            outcome = "failed"
    return dict(category=category, severity=severity, outcome=outcome, error_code=error_code)



class DiagnosticHarnessReferenceV1(BaseModel):
    model_config = ConfigDict(extra="forbid", strict=True)
    operation_id: Annotated[str, Field(pattern=HARNESS_OPERATION_ID_PATTERN)]
    workflow: HarnessWorkflow
    stage: HarnessStage
    trace_id: Annotated[str, Field(pattern=r"^[A-Za-z0-9:._-]{1,160}$")] | None = None
    attempt_id: Annotated[str, Field(pattern=r"^[A-Za-z0-9:._-]{1,160}$")] | None = None
    attempt_index: Annotated[int, Field(ge=1)] | None = None
    phase: HarnessAttemptPhase | None = None
    attempt_status: HarnessAttemptStatus | None = None

    @model_validator(mode="after")
    def coherent_attempt(self):
        fields = (self.attempt_id, self.attempt_index, self.phase, self.attempt_status)
        if any(value is not None for value in fields) and (self.trace_id is None or any(value is None for value in fields)):
            raise ValueError("diagnostic_attempt_reference_incomplete")
        return self

    # A reference is an index hint; clients must resolve canonical records for truth.


class DiagnosticResourceReferenceV1(BaseModel):
    model_config = ConfigDict(extra="forbid", strict=True)
    resource_type: Literal["persona"]
    resource_id: Annotated[str, Field(pattern=r"^[A-Za-z0-9:._-]{1,160}$")]
    revision: Annotated[int, Field(ge=0)] | None = None


class DiagnosticEventV1(BaseModel):
    model_config = ConfigDict(extra="forbid", strict=True)
    schema_version: Literal["diagnostic-event-v1"] = "diagnostic-event-v1"
    event_id: Identity
    source: Literal["server", "browser", "desktop"]
    name: Literal["decode_failed", "page_entered", "page_left", "request_started", "response_headers", "request_finished", "request_failed", "request_cancelled", "lifecycle_started", "lifecycle_stopped", "harness_reference", "resource_reference", "provider_started", "provider_finished", "provider_failed", "provider_attempt_started", "provider_attempt_finished", "provider_attempt_failed", "tool_started", "tool_finished", "tool_failed", "tool_unknown", "action_started", "action_finished", "action_failed", "action_cancelled", "desktop_started", "sidecar_spawned", "sidecar_ready", "sidecar_startup_failed", "sidecar_exited", "desktop_shutdown_requested", "sidecar_stopped", "sidecar_shutdown_unknown", "desktop_stopped"]
    page_path: DiagnosticPagePath | None = None
    desktop_metric: DiagnosticDesktopMetricV1 | None = None
    action_name: Literal["json_import_read_persona", "json_import_read_scene", "json_export_handoff", "settings_save", "vault_create", "vault_unlock", "vault_lock", "vault_load_secrets", "vault_save_secrets", "vault_clear_secrets"] | None = None
    span_id: Identity | None = None
    parent_span_id: Identity | None = None
    tool_metric: DiagnosticToolMetricV1 | None = None
    provider_metric: DiagnosticProviderMetricV1 | None = None
    harness: DiagnosticHarnessReferenceV1 | None = None
    resource: DiagnosticResourceReferenceV1 | None = None
    category: Literal["transport", "lifecycle", "harness", "resource", "provider", "tool", "action", "desktop", "page", "decode"]
    severity: Literal["info", "warning", "error"]
    outcome: Literal["started", "headers_received", "completed", "failed", "cancelled", "observed", "unknown"]
    error_code: Literal["transport_failure", "http_error", "cancelled", "provider_failure", "tool_failure", "action_failure", "desktop_failure", "decode_failure"] | None

    @model_validator(mode="before")
    @classmethod
    def reviewed_classification(cls, value):
        if isinstance(value, dict) and isinstance(value.get("name"), str) and value.get("name") in DIAGNOSTIC_EVENT_CATALOG:
            expected = classify_diagnostic(value["name"], value.get("status_code"))
            if any(key in value and value[key] != item for key, item in expected.items()):
                raise ValueError("diagnostic_classification_mismatch")
            value = {**value, **expected}
        return value

    timestamp: Annotated[str, Field(max_length=40)]
    request_id: Identity | None = None
    client_instance_id: Identity | None = None
    page_view_id: Identity | None = None
    flow_id: Identity | None = None
    action_id: Identity | None = None
    duration_ms: Annotated[float, Field(ge=0, allow_inf_nan=False)] | None = None
    status_code: Annotated[int, Field(ge=100, le=599)] | None = None
    method: Literal["GET", "POST", "PUT", "PATCH", "DELETE", "OPTIONS", "HEAD"] | None = None
    # Reviewed route templates only; raw paths, query strings and bodies are excluded.
    route: Annotated[str, Field(max_length=160, pattern=r"^/[a-zA-Z0-9_/{\}-]*$")] | None = None
